### ISSUE TITLE
bug in revised AMOEBA plugin code for Vdw interactions

### ISSUE THREAD

**Issue by jayponder** [openmm/openmm#2830]
There is a bug in the recently revised AMOEBA plugin code for Vdw interactions. This would be the code updated under the merged/closed pull request "Added new features to AmoebaVdwForce #2765".

The problem occurs for a system having a number of atoms that is an exact multiple of 32. For those systems, there is an attempt to access an undefined atom off the end of an atom block. This was patched in the earlier pull request we put up some time ago ("AMOEBA Update, 8 May 2020 #2674"), but the patch did not make it into the final revised vdw version.

Code from the original pull request that works correctly can be found at:

https://github.com/jayponder/openmm/blob/AMOEBA-Update/plugins/amoeba/platforms/cuda/src/kernels/amoebaVdwForce3.cu

In particular, note the "if clause" shown below that is added around the bulk of the code in the above link from the original pull request. This if clause is missing or was removed in the revised current version.

if (atom2 < NUM_ATOMS) {
.
.
}

There may be other ways to solve this, perhaps deeper inside OpenMM, but the above patch seems to be a general solution in our hands.

Also, in the same section of the revised current code, the variable "tempForce" is not used and can be removed.

**Comment by peastman** [comment]
Do you have an example that demonstrates the problem?  I'm looking through the code, but I don't see how it could happen.  The corresponding code in the current version is in `amoebaVdwForce2.cu`.  The only memory access in that block is where it loads the parameters from the table:

https://github.com/openmm/openmm/blob/b3d9846981bb47e3eb937d552e90b5db960b5a1b/plugins/amoeba/platforms/cuda/src/kernels/amoebaVdwForce2.cu#L8

All the arrays for parameters have their size padded to a multiple of 32, with the extra elements set to 0.  So if `atom2 >= NUM_ATOMS`, then `atomType2` will be 0.  That shouldn't cause an illegal access.  Also in that case `isExcluded` will be set, so no force or energy contribution will be added:

https://github.com/openmm/openmm/blob/b3d9846981bb47e3eb937d552e90b5db960b5a1b/plugins/amoeba/platforms/cuda/src/kernels/amoebaVdwForce2.cu#L62-L63

**Comment by zhi-wang** [comment]
In this corner case where `NUM_ATOMS == PADDED_NUM_ATOMS` and `atom2 == NUM_ATOMS`, the attempt to access element in an array of length `PADDED_NUM_ATOMS` via:
```C
float value = array[atom2];
```
is illegal.

We haven't seen similar problems if `NUM_ATOMS < PADDED_NUM_ATOMS`.

**Comment by peastman** [comment]
Can you point me to the line where that's happening?  `atom2` should always be less than `PADDED_NUM_ATOMS`, so the case you describe should never happen.

**Comment by zhi-wang** [comment]
We saw crashes and `atom2 == NUM_ATOMS == PADDED_NUM_ATOMS` in two systems. One had 6432 atoms, the other one had 1920 atoms. What we did was to add a print statement in "vdw2.cu":
```C
if (atoms2 == NUM_ATOMS) printf(" atom2 = %d atomType2 = %d\n", atom2, atomType2);
float2 pairSigmaEpsilon = sigmaEpsilon[atomType1+atomType2*NUM_TYPES];
```

As a matter of fact, we also saw the printing from CUDA kernel if `NUM_ATOMS` was not a multiple of 32. Since `atomType` was allocated longer than `NUM_ATOMS` in this case, reading from `atomType[NUM_ATOMS]` was okay.

I am not familiar with the underlying CUDA algorithms in OpenMM, and the fact that the CUDA source code was generated on the fly makes it more confusing to me. I don't know exactly which line caused it. All I know is, the `interactingAtoms` array of the neighbor list was "padded" by `NUM_ATOMS`.
https://github.com/openmm/openmm/blob/b3d9846981bb47e3eb937d552e90b5db960b5a1b/platforms/cuda/src/kernels/findInteractingBlocks.cu#L369

My best guess is this may be the origin of the value that `atom2` was set to. I don't know if there is any safe guard in the energy evaluation kernels to check or update the value of `atom2`, or how the safe guard was applied in the kernel.

Thank you.

**Comment by peastman** [comment]
Can you provide a test case that demonstrates the problem?  If I can reproduce it, I can analyze it and figure out what's going on.  If I can't reproduce it, it's very difficult for me to do anything.

**Comment by jayponder** [comment]
Sure I'll work up a PDB file and Python script for you as a test case (we do stuff out of Tinker, but that's probably not what you want). Our feeling is that any system with a total number of atoms that is an exact multiple of 32 will fail. So I'll send 32 water molecules in a periodic box, or something similar.

**Comment by peastman** [comment]
Thanks for the excellent test case!  It took me a little while to figure out exactly what's going on.  The neighbor list consists of blocks of 32 atoms.  Often there will be a partially filled block at the end, so it marks the unused atoms by setting them to `NUM_ATOMS`, which in your example also happens to equal `PADDED_NUM_ATOMS`.

In the kernel to compute nonbonded interactions, it checks for that before loading atom parameters:

https://github.com/openmm/openmm/blob/c4b9505a610f3c59c55de2a685b7258bcabcd1a4/platforms/cuda/src/kernels/nonbonded.cu#L399-L422

In the `if` block, which is for real atoms, it loads the position from memory and then the symbol `LOAD_LOCAL_PARAMETERS_FROM_GLOBAL` gets replaced with code to load any other parameters.  The `else` block, which is for padding atoms, sets the position to 0 but fails to clear other parameters.  If the very first block a particular warp processes happens to be an incomplete one, that memory will never have been initialized.  Even that would usually not cause any problems: whatever calculation gets done based on the uninitialized memory, the result gets thrown out and doesn't effect anything.  But in this case it's being used to compute an offset into a table, and that can lead to an invalid read.

I believe the correct fix is to make the `else` block initialize all parameters to 0 for the padding atoms.

**Comment by zhi-wang** [comment]
Thank you very much!

**Comment by peastman** [comment]
The fix is in #2837.
